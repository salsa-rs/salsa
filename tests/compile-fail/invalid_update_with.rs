#[derive(salsa::Update)]
struct S2 {
    #[update(unsafe(with(my_wrong_update)))]
    bad: i32,
    #[update(unsafe(with(my_wrong_update2)))]
    bad2: i32,
    #[update(unsafe(with(my_wrong_update3)))]
    bad3: i32,
    #[update(unsafe(with(true)))]
    bad4: &'static str,
}

fn my_wrong_update() {}
fn my_wrong_update2(_: (), _: ()) -> bool {
    true
}
fn my_wrong_update3(_: *mut i32, _: i32) -> () {}

fn main() {}
