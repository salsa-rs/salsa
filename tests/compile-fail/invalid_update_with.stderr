error[E0308]: mismatched types
 --> tests/compile-fail/invalid_update_with.rs:3:26
  |
3 |     #[update(unsafe(with(my_wrong_update)))]
  |                     ---- ^^^^^^^^^^^^^^^ incorrect number of function parameters
  |                     |
  |                     expected due to this
  |
  = note: expected fn pointer `unsafe fn(*mut i32, i32) -> bool`
                found fn item `fn() -> () {my_wrong_update}`

error[E0308]: mismatched types
 --> tests/compile-fail/invalid_update_with.rs:5:26
  |
5 |     #[update(unsafe(with(my_wrong_update2)))]
  |                     ---- ^^^^^^^^^^^^^^^^ expected fn pointer, found fn item
  |                     |
  |                     expected due to this
  |
  = note: expected fn pointer `unsafe fn(*mut i32, i32) -> bool`
                found fn item `fn((), ()) -> bool {my_wrong_update2}`

error[E0308]: mismatched types
 --> tests/compile-fail/invalid_update_with.rs:7:26
  |
7 |     #[update(unsafe(with(my_wrong_update3)))]
  |                     ---- ^^^^^^^^^^^^^^^^ expected fn pointer, found fn item
  |                     |
  |                     expected due to this
  |
  = note: expected fn pointer `unsafe fn(*mut i32, i32) -> bool`
                found fn item `fn(*mut i32, i32) -> () {my_wrong_update3}`

error[E0308]: mismatched types
 --> tests/compile-fail/invalid_update_with.rs:9:26
  |
9 |     #[update(unsafe(with(true)))]
  |                     ---- ^^^^ expected fn pointer, found `bool`
  |                     |
  |                     expected due to this
  |
  = note: expected fn pointer `unsafe fn(*mut &'static str, &'static str) -> bool`
                   found type `bool`
